[Custom Map] Single Player Maze

coreyja opened the pull request:
Took a few weeks on stream but we got a Single Player Maze custom map all working nicely now!

This map works best of a fairly large board, I've been doing 25X25 and 50x50 locally

It starts by drawing a 7x7 maze in the center of the board, with your snake in the bottom left corner and a single food in the top right corner

The goal is for your snake to reach the food! When you reach the food you "complete" that level, and the map generates a brand new maze for you to solve! This time the height and width will be 1 greater than last time, so 11x11 on the second level (`level 1` due to zero indexed levels)

If you solve all the levels up to the size of the board, the maze will stop growing and stay the same size going forward.

Once you've solved the largest board size a couple times (currently 5) it turns on 'evil mode' where as you get close to food it will move away from you, forcing the game to end!


[Old] GIF CLI Demo: For more recent demos see gifs down thread

![2022-06-12 14 56 08](https://user-images.githubusercontent.com/964771/173249687-4e4c2275-7d5a-4ce5-aced-321a010da016.gif)

coreyja commented on maps/coreyja_maze.go:
```diff
@@ -0,0 +1,447 @@
+}
+
+func (m CoreyjaMazeMap) ID() string {
+	return "coreyja_maze"
```
I can get more creative with a name too if we want!

coreyja commented:
Ok some updates!

Added a version number! Set it to `0` right now, but can bump it to `1` if it makes a difference

I changed the maze generation to make less 'cuts' in the walls, which led to much more interesting mazes!

I also changed the food spawn to be a random position, which also leads to much more interesting games!

With both those changes (and a large enough board) we get into situations where its not possible to get to the food before running out of health. I can fix that if I want but its a kinda nice way of limiting game length, but probably won't work well on small boards. Still might need an artificial cap to prevent infinite games

Heres a new demo video!

https://user-images.githubusercontent.com/964771/174351860-8fb4993c-ab2e-4564-9734-79894bf0471b.mp4

bvanvugt commented:
@coreyja Oooo yes, this is really great. I am genuinely surprised how much impact the random food location has on making it fun/mesmerizing to watch.

Just for transparency - I'd like to get this live as quick as we can, and launch some challenges against it. There are two blockers I'm aware of:
* We need to address the infinite turns problem
* I want @robbles to take a look at the code before we merge/ship (he's back from time-off this week)

Re: Games running indefinitely, I'm almost of the mind that maps should also have the control to indicate a game has ended. One way to do this could be allowing it to eliminate snakes, via some sort of 'Eliminated by Map' cause. Another might be to allow maps to indicate game over inside the UpdateBoard func... 🤔 curious what you think.

A temporary idea - for version 0, after N (10?) successful mazes you could surround the snake with hazards and force it to eliminate itself. That would at the very least prevent run-away games.

coreyja commented:
> I'm almost of the mind that maps should also have the control to indicate a game has ended

Oh I hadn't thought of that! Its an interesting idea!
I think I could fake it by just removing the snake from the board, or putting them inside a wall or something like that!

> A temporary idea - for version 0, after N (10?) successful mazes you could surround the snake with hazards and force it to eliminate itself. That would at the very least prevent run-away games.

Ya I think this makes sense! I'll try to throw something like this in there!

robbles commented:
Hey @coreyja! We had a quick chat this morning and decided on a few next steps for this map:
- It needs a manual endgame condition for now, which I think you're already looking at. Pretty much anything that guarantees the game will eventually end, whether surrounding the snake with hazards, not spawning more food, etc.
- It looks like there's some linting issues, plus some missing metadata that the unit tests expect. We recently added support for specifying a version, minimum and maximum players, and allowed board sizes. Let me know if you have questions around what the tests expect - we may need to tweak them a bit to accommodate this map.
- When running against the web-based board, it looked like some of the hazards were stacking up in a way that's probably not visible from the terminal UI. It doesn't seem to break anything, but probably should be avoided so that it looks good on the board UI.  See below:

<img width="989" alt="Screen Shot 2022-06-23 at 9 42 40 AM" src="https://user-images.githubusercontent.com/92927/175382716-351fd499-4e86-48ca-b070-360afe3028b7.png">

The production engine is currently (sort of arbitrarily) limited to board sizes of 25 x 25, but it looks to me like this map will work OK at that size?

After those changes, I think we can merge this and have it in an "open beta" for anyone to play test locally!

We're hoping to work this map into a couple of new challenges soon - are you interested in coming up with the configuration and win conditions for those? e.g.
- 1 snake, 25 x 25, win if you survive to turn N
- 1 snake, 25 x 25, constrictor mode, win if you survive to turn N
- etc.

Still thinking about Brad's idea above for adding the ability to end the game explicitly from a map, but I think that's something we can work into a version 2.

coreyja commented:
OK! I think this probably ready to go!

We fixed the duplicate hazards and got the meta data and tests working!

Added an `evil mode` that takes effect once you've been at the largest possible maze for currently 5 turns. After this whenever you get close to the food it picks a new location! This forces the game to end since you can't get the food and the snake will eventually starve

One thing I forgot about till now is the Challenges!

I think a good challenge would probably be something like:
25x25 board survive till turn 400, which seems to be a good turn count. From watching some with my snake it seems like at turn 400 'evil mode' shouldn't have kicked in yet, but is getting close to that point. And means you at least solved 4 mazes (likely more) so feels like a pretty good challenge!

coreyja commented:
Oh I'll leave some thoughts on the Map interface here too!

1. The `editor` interface instead of modifying the `BoardState` makes it hard to 'read your own writes' which would have been helpful for me. To work around this I created a clone of the `BoardState` and I mirrored the calls to the `Editor` with changes to the BoardState. I passed around this board state so I could query it for things like hazard/wall position later on. [I actually have two BoardStates that I pass around, one which is the size of the maze and one that is the size of the map. But that's not super related to the Map interface.]
	- Link: https://github.com/BattlesnakeOfficial/rules/pull/81/files#diff-33aeca5eeb757a00a8e248fa2fbfce5f9adaf99e4578d96437c48d80350ace9aR90-R91
2. Having the ability to set a minimum size board that the map supports. Anything smaller than a 7x7 doesn't work super well for the maze, and I have it erroring on setup if we try a board that small. It would be nice to be able to represent that in the meta data

coreyja commented:
Oh I would also recommend a "Squash and Merge" strategy here, or I can rebase all this into 1 commit if that's easier!

There are lots of small commits here, and a couple times I cherry picked and reverted the CLI -> Board integration commit lol

coreyja commented:
> For the minimum board size - are you looking to express something like >7, >7, as opposed to unlimited or a list of specific sizes? 

Ya that is what I was thinking!

>  I think we could support that in the near future, but for now maybe just put a specific list of sizes in there? Something like FixedSizes(Dimensions{7, 7}, Dimensions{11,11}, Dimensions{19,19}, Dimensions{25,25}) would cover a reasonable range of sizes that the engine currently supports.

Yup that feels like a reasonable thing to do in the short term! Will try to get this updated with that this evening

robbles commented:
One of my snakes with a "stay small" strategy seems to have found an interesting way to kill time on the first round of the maze:
![2022-06-29 17 01 44](https://user-images.githubusercontent.com/92927/176565681-4bbff28d-7aa4-49ac-b53b-98df584d8dd3.gif)
I think maybe the tail length should be 1 longer for all rounds, to prevent this?

coreyja commented:
Oh wow thats a fun one!

Ya I think starting snake length at 3 makes sense then! There was another weird issue we saw with 1 length too, do defaulting to 3 makes sense!

coreyja commented:
Added the two things I wanted to push!

Added 1 more to the length at every level, and also hard coded the list of supported sizes to all the ones the web UI supports now!

coreyja commented:
Re: Map Interface

> Editor.ReadTile or Editor.IsTileOccupied?

Ya I think something that can read a position and return any food/hazards/snakes in that position would go long way to not needing to have a 'mirror' BoardState

Without thinking too hard about the implementation, I wonder if there is a way to make the interface work a tiny bit easier with a mirror board state? Like maybe a `Editor.CopyFromBoardState` that clears the board and the copies things over from the BoardState? Basically just a helper to make the mirror board state stuff a bit less manual? 🤷

robbles reviewed (approved):
Thanks for updating the metadata! This works well for me when testing locally.

I think it's probably worth doing some reorganization of the helpers and debugging functions, and it could do with some tests, but I'm good with merging for now since we'll probably tweak it some more before it's rolled out to the production engine.  That way anyone can playtest it locally.
